File: tasks/translation/output/Qwen/Qwen2.5-32B-Instruct-AWQ/codenet/direct/Java/Python/s584572995_commented.py
mod = 10**9 + 7

def modpow(x, y):
    if y == 0:
        return 1
    if y % 2 != 0:
        return x * modpow(x, y-1) % mod
    tmp = modpow(x, y // 2)
    return tmp * tmp % mod

def rec(v, parent):
    res = 1
    for next in to[v]:
        if next == parent:
            continue
        res += rec(next, v)
    dp[v] = res
    return res

class FastScanner:
    def __init__(self, in_stream):
        self.reader = in_stream
        self.tokenizer = None

    def next(self):
        if self.tokenizer is None or not self.tokenizer.hasMoreTokens():
            self.tokenizer = StringTokenizer(self.reader.readline())
        return self.tokenizer.nextToken()

    def nextLine(self):
        if self.tokenizer is None or not self.tokenizer.hasMoreTokens():
            return self.reader.readline()
        return self.tokenizer.nextToken("\n")

    def nextLong(self):
        return int(self.next())

    def nextInt(self):
        return int(self.next())

    def nextDouble(self):
        return float(self.next())

    def nextIntArray(self, n):
        return [self.nextInt() for _ in range(n)]

    def nextLongArray(self, n):
        return [self.nextLong() for _ in range(n)]

class StringTokenizer:
    def __init__(self, line):
        self.tokens = line.split()
        self.index = 0

    def hasMoreTokens(self):
        return self.index < len(self.tokens)

    def nextToken(self, delimiter=None):
        if delimiter is None:
            token = self.tokens[self.index]
            self.index += 1
            return token
        else:
            return self.tokens[self.index].split(delimiter)[0]

import sys
fs = FastScanner(sys.stdin)
N = fs.nextInt()
to = [[] for _ in range(N)]
edges = []
for _ in range(N-1):
    A, B = fs.nextInt()-1, fs.nextInt()-1
    to[A].append(B)
    to[B].append(A)
    edges.append((B, A))

dp = [0] * N
rec(0, -1)

mulsum = 0
for e in edges:
    candidate1, candidate2 = e
    child = candidate1 if dp[candidate1] < dp[candidate2] else candidate2
    a = dp[child]
    b = N - a
    mul = (modpow(2, a) - 1) * (modpow(2, b) - 1) % mod
    mulsum += mul
    mulsum %= mod

cases = modpow(2, N)
PN = (mulsum + (cases - 1 + mod) % mod) % mod
BN = N * modpow(2, N-1) % mod
WN = (PN - BN + mod) % mod
ans = WN * modpow(cases, mod-2) % mod

print(ans)
# End of Code

File: tasks/translation/output/Qwen/QwQ-32B-AWQ/codenet/direct/multiline/uncomment/direct/Java/C++/s584572995_commented.cpp
#include <iostream>
#include <vector>
#include <string>
#include <sstream>
using namespace std;

const int mod = 1e9 +7;

struct FastScanner {
    istream& in;
    string buffer;
    stringstream ss;
    vector<string> tokens;
    size_t ptr;
public:
    FastScanner(istream& _in) : in(_in), ptr(0) {}
    
    string next() {
        loadTokensIfEmpty();
        if (ptr >= tokens.size()) {
            loadNextLine();
            loadTokensIfEmpty();
        }
        return tokens[ptr++];
    }
    
    string nextLine() {
        loadNextLine();
        return buffer;
    }
    
    int nextInt() { return stoi(next()); }
    long long nextLong() { return stoll(next()); }
    double nextDouble() { return stod(next()); }
    
    vector<int> nextIntArray(int n) {
        vector<int> arr(n);
        for(int i=0; i<n; i++) arr[i] = nextInt();
        return arr;
    }
    
    vector<long long> nextLongArray(int n) {
        vector<long long> arr(n);
        for(int i=0; i<n; i++) arr[i] = nextLong();
        return arr;
    }
    
private:
    void loadNextLine() {
        if (!getline(in, buffer)) {
            throw runtime_error("Error reading line.");
        }
        ss.clear();
        ss.str(buffer);
        tokens.clear();
        string token;
        while (ss >> token) {
            tokens.push_back(token);
        }
        ptr = 0;
    }
    
    void loadTokensIfEmpty() {
        if (tokens.empty()) {
            loadNextLine();
        }
    }
};

int rec(int v, int parent, vector<int>& dp, const vector<vector<int>>& to) {
    int res = 1;
    for(int next : to[v]) {
        if(next == parent) continue;
        res += rec(next, v, dp, to);
    }
    dp[v] = res;
    return res;
}

long long modpow(long long x, long long y) {
    long long res = 1;
    while(y > 0) {
        if(y % 2 == 1) res = res * x % mod;
        x = x * x % mod;
        y /= 2;
    }
    return res;
}

int main() {
    FastScanner fs(cin);
    int N = fs.nextInt();
    vector<vector<int>> to(N);
    vector<pair<int, int>> edges;
    for(int i=0; i<N-1; i++) {
        int A = fs.nextInt()-1, B = fs.nextInt()-1;
        to[A].push_back(B);
        to[B].push_back(A);
        edges.push_back({B, A});
    }
    vector<int> dp(N);
    rec(0, -1, dp, to);
    
    long long mulsum = 0;
    for(auto& e : edges) {
        int candidate1 = e.first;
        int candidate2 = e.second;
        int child;
        if(dp[candidate1] > dp[candidate2]) {
            child = candidate2;
        } else {
            child = candidate1;
        }
        long long a = dp[child];
        long long b = N - a;
        long long term = ( (modpow(2, a) -1) % mod ) * ( (modpow(2, b) -1) % mod ) % mod;
        mulsum = (mulsum + term) % mod;
    }
    
    long long cases = modpow(2, N);
    long long PN = (mulsum + (cases -1 + mod) % mod) % mod;
    long long BN = (N * modpow(2, N-1)) % mod;
    long long WN = (PN - BN + mod) % mod;
    long long ans = WN * modpow(cases, mod-2) % mod;
    
    cout << ans << endl;
    return 0;
}
// End of Code
